File: app/src/main/java/org/cosmic/ide/adapter/ProjectAdapter.java
package org.cosmic.ide.adapter;

import android.view.View;
import android.view.ViewGroup;
import android.view.LayoutInflater; 
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.DiffUtil;

import org.cosmic.ide.R;
import org.cosmic.ide.project.JavaProject;

import java.util.ArrayList;
import java.util.List;

public class ProjectAdapter extends RecyclerView.Adapter<ProjectAdapter.ViewHolder> {

    public interface OnProjectSelectedListener {
        void onProjectSelect(JavaProject project);
    }

    public interface OnProjectLongClickedListener {
        boolean onLongClicked(JavaProject project);
    }

    private final List<JavaProject> mProjects = new ArrayList<>();
    public OnProjectSelectedListener onProjectSelectedListener;
    public OnProjectLongClickedListener onProjectLongClickedListener; 

    public void setOnProjectSelectedListener(OnProjectSelectedListener onProjectSelectedListener) {
        this.onProjectSelectedListener = onProjectSelectedListener;
    }
    
    public void setOnProjectLongClickedListener(OnProjectLongClickedListener onProjectLongClickedListener) {
        this.onProjectLongClickedListener = onProjectLongClickedListener;
    }

    public void submitList(@NonNull List<JavaProject> projects) {
         var diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                 return mProjects.size();
            }

            @Override
            public int getNewListSize() {
                return projects.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return mProjects.get(oldItemPosition).equals(projects.get(newItemPosition));
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                 return mProjects.get(oldItemPosition).equals(projects.get(newItemPosition));
            }
         });
         mProjects.clear();
         mProjects.addAll(projects);
         diffResult.dispatchUpdatesTo(this);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        var root = LayoutInflater.from(parent.getContext()).inflate(R.layout.project_item, parent, false);
        final var holder = new ViewHolder(root);
        return holder;
    }

    @NonNull
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(mProjects.get(position));
        holder.background.setOnClickListener(v -> {
            if (onProjectSelectedListener != null && position != RecyclerView.NO_POSITION) {
                onProjectSelectedListener.onProjectSelect(mProjects.get(position));
            }
        });
        holder.background.setOnLongClickListener(v -> {
            if(onProjectLongClickedListener != null) {
                if(position != RecyclerView.NO_POSITION) {
                    return onProjectLongClickedListener.onLongClicked(mProjects.get(position));
                }
            }
            return false;
        });
    }

    @Override
    public int getItemCount() {
        return mProjects.size(); 
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView title;
        private final TextView path;
        public final View background;

        public ViewHolder(View view) {
            super(view);
            title = view.findViewById(R.id.project_title);
            path = view.findViewById(R.id.project_path);
            background = view.findViewById(R.id.background);
        }

        public void bind(JavaProject project) {
            title.setText(project.getProjectName());
            path.setText(project.getProjectDirPath());
        }
    }
}

File: app/src/main/kotlin/org/cosmicide/adapter/ProjectAdapter.kt
/*
 * This file is part of Cosmic IDE.
 * Cosmic IDE is a free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * Cosmic IDE is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with Cosmic IDE. If not, see <https://www.gnu.org/licenses/>.
 */

/*
 * This file is part of Cosmic IDE.
 * Cosmic IDE is a free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * Cosmic IDE is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with Cosmic IDE. If not, see <https://www.gnu.org/licenses/>.
 */

package org.cosmicide.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import org.cosmicide.databinding.ProjectItemBinding
import org.cosmicide.project.Project
import kotlin.properties.Delegates

/**
 * Adapter for displaying a list of [Project] objects in a [RecyclerView].
 *
 * @param listener Listener for handling user interactions with the items in the list.
 */
class ProjectAdapter(private val listener: OnProjectEventListener) :
    RecyclerView.Adapter<ProjectAdapter.ViewHolder>() {

    /**
     * List of [Project] objects to be displayed in the [RecyclerView].
     */
    private var projects: List<Project> by Delegates.observable(emptyList()) { _, oldList, newList ->
        DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize() = oldList.size
            override fun getNewListSize() = newList.size
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                oldList[oldItemPosition] == newList[newItemPosition]

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                oldList[oldItemPosition] == newList[newItemPosition]
        }).dispatchUpdatesTo(this)
    }

    /**
     * Interface for handling user interactions with the items in the list.
     */
    interface OnProjectEventListener {
        fun onProjectClicked(project: Project)
        fun onProjectLongClicked(project: Project, v: View): Boolean
    }

    /**
     * Sets the list of projects to be displayed in the [RecyclerView].
     *
     * @param projects The list of [Project] objects.
     */
    fun submitList(projects: List<Project>) {
        this.projects = projects
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
        ViewHolder(
            ProjectItemBinding.inflate(LayoutInflater.from(parent.context), parent, false),
            listener
        )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(projects[position])
    }

    override fun getItemCount() = projects.size

    /**
     * ViewHolder for displaying a single [Project] object in the [RecyclerView].
     *
     * @param itemBinding The [ProjectItemBinding] object for the item view.
     * @param listener The [OnProjectEventListener] for handling user interactions with the item view.
     */
    inner class ViewHolder(
        itemBinding: ProjectItemBinding,
        private val listener: OnProjectEventListener
    ) : BindableViewHolder<Project, ProjectItemBinding>(itemBinding) {

        override fun bind(data: Project) {
            binding.projectTitle.text = data.name
            binding.projectPath.text = data.root.absolutePath
            binding.root.setOnClickListener { listener.onProjectClicked(data) }
            binding.root.setOnLongClickListener {
                listener.onProjectLongClicked(
                    data,
                    binding.root
                )
            }
        }
    }
}